book/src/repair-service.md

```diff
@@ -0,0 +1,50 @@
+# Repair Service
+
+The RepairService is in charge of retrieving missing blobs that failed to be delivered by primary communication protocols like Avalanche. It is in charge of managing the protocols described below in the `Repair Protocols` section below.  
+
+# Repair Protocols
+
+The repair protocol makes best attempts to progress the forking structure of Blocktree. 
+
+The different protocol strategies:
+
+1. Fast Repair:
+    Problem: Validators that find themselves behind the network by an entire epoch struggle/fail to catch up because they do not have a leader schedule for future epochs. If nodes were to blindly accept repair blobs in these future epochs, this exposes nodes to spam.
+
+    Solution: This part of the repair protocol is the primary mechanism by which new nodes joining the cluster catch up after loading a snapshot. This protocol works in a "forward" fashion, so validators can verify every blob that they receive against a known leader schedule.
+
+    Each validator advertises in gossip:
+        * Current root
+        * The set of all completed slots in the confirmed epochs (an epoch that was calculated based on a bank <= current root) past the current root
+    
+    Observers of this gossip message with higher epochs (repairmen) send blobs to catch the lagging node up with the rest of the cluster. The repairmen are responsible for sending the slots within the epochs that are confrimed by the advertised `root` in gossip. The repairmen divide the responsibility of sending each of the missing slots in these epochs based on a random seed (simple blob.index iteration by N, seeded with the repairman's node_id). Ideally, each repairman in an N node network (N nodes whose epochs are higher than that of the repairee) sends 1/N of the missing blobs. Both data and coding blobs for missing slots are sent. Repairmen do not send blobs again to the same validator until they see the message in gossip updated, at which point they perform another iteration of this protocol.
```

Comment:
@rob-solana, not sure about using a different random seed per node, seems like you might get a lot of overlap. If we had a way to totally order the nodes in gossip, we might get better distribution of responsibility (Maybe have nodes pick random numbers in a range and gossip those to establish a total ordering?)

Reply:
what if that's still the whole network?

you're saying the distribution of the last digits of leader_id can't be uniform, and therefore we'll have a lot of duplicate send()s?

Reply:
Right, if you have `n` nodes and `n` buckets, and each node is picking a random bucket, then nodes will pick a lot of duplicate buckets

Reply:
what's the fix?  it doesn't need to be perfect, assuming erasure.

How about seeding a prng with the node_id and sending rng.gen(slot_max_idx) for slot_max_idx/num nodes?

Sounds like the same thing to me...

Reply:
oh, yeah: it'd be cool if the send_to() loop waited for some fraction of slot_ms/max_slot_idx between each blob ;)

Reply:
Not sure about the fix, but was considering the tradeoff if instead of a random seed per validator, each validator orders the peers in gossip that are at least an epoch ahead of the repairee, and used that ordering to determine responsibility for each blob.

What is the send_to() loop you're referring to? you mean when the repairment are sending the blobs to the repairee?

Reply:
yeah, that send_to() loop.

on "orders the peers in gossip": what value would determine the order and how many will be repairmen?

Reply:
Everyone with an advertised root in gossip greater that's at least one epoch ahead of the repairee will be a repairman? I don't think the order matters as much as long as everybody agrees on the ordering, so maybe based on Pubkey?

Reply:
ok, that should work nicely